File: java/src/test/java/libldt3/test/TestF003CornerCases.java
package libldt3.test;

import libldt3.LdtConstants;
import libldt3.LdtReader;
import libldt3.model.saetze.Auftrag;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.time.LocalDate;
import java.time.Year;
import java.time.YearMonth;

public class TestF003CornerCases {

    @Test
    public void testWithYearOnly() throws IOException {
        var reader = new LdtReader(LdtConstants.Mode.RELAXED);

        var read = reader.read(getClass().getResource("/person-19900000.ldt").getPath());

        Assertions.assertNotNull(read);
        Assertions.assertEquals(1, read.size());
        Assertions.assertNotNull(read.get(0));
        Assertions.assertInstanceOf(Auftrag.class, read.get(0));
        Assertions.assertNotNull(((Auftrag)read.get(0)).patient);
        Assertions.assertNotNull(((Auftrag)read.get(0)).patient.person);
        Assertions.assertNotNull(((Auftrag)read.get(0)).patient.person.geburtsdatum);
        Assertions.assertInstanceOf(Year.class, ((Auftrag)read.get(0)).patient.person.geburtsdatum);
        Assertions.assertEquals(Year.of(1900), ((Auftrag)read.get(0)).patient.person.geburtsdatum);


    }

    @Test
    public void testWithYearAnyMonth() throws IOException {
        var reader = new LdtReader(LdtConstants.Mode.RELAXED);

        var read = reader.read(getClass().getResource("/person-19900200.ldt").getPath());

        Assertions.assertNotNull(read);
        Assertions.assertEquals(1, read.size());
        Assertions.assertNotNull(read.get(0));
        Assertions.assertInstanceOf(Auftrag.class, read.get(0));
        Assertions.assertNotNull(((Auftrag) read.get(0)).patient);
        Assertions.assertNotNull(((Auftrag) read.get(0)).patient.person);
        Assertions.assertNotNull(((Auftrag) read.get(0)).patient.person.geburtsdatum);
        Assertions.assertInstanceOf(YearMonth.class, ((Auftrag) read.get(0)).patient.person.geburtsdatum);
        Assertions.assertEquals(YearMonth.of(1900, 2), ((Auftrag) read.get(0)).patient.person.geburtsdatum);
    }

    @Test
    public void testWithFullDate() throws IOException {
        var reader = new LdtReader(LdtConstants.Mode.RELAXED);

        var read = reader.read(getClass().getResource("/person-19900204.ldt").getPath());

        Assertions.assertNotNull(read);
        Assertions.assertEquals(1, read.size());
        Assertions.assertNotNull(read.get(0));
        Assertions.assertInstanceOf(Auftrag.class, read.get(0));
        Assertions.assertNotNull(((Auftrag) read.get(0)).patient);
        Assertions.assertNotNull(((Auftrag) read.get(0)).patient.person);
        Assertions.assertNotNull(((Auftrag) read.get(0)).patient.person.geburtsdatum);
        Assertions.assertInstanceOf(LocalDate.class, ((Auftrag) read.get(0)).patient.person.geburtsdatum);
        Assertions.assertEquals(LocalDate.of(1900, 2, 4), ((Auftrag) read.get(0)).patient.person.geburtsdatum);
    }
}

File: cs/libldt3.tests/Tests/libldt3/test/TestF003CornerCases.cs
using System;
using System.IO;
using NUnit.Framework;
using NodaTime;
using libldt3.model;

namespace libldt3
{
    namespace test
    {
        class TestF003CornerCases
        {
            private static string getTempFile(String prefix)
            {
                return Path.Combine(Path.GetTempPath(), Path.GetRandomFileName(), ".ldt");
            }

            [Test]
            public void testWithYearOnly()
            {
                var reader = new LdtReader(LdtConstants.Mode.RELAXED);
                string inFile = Path.Combine(TestContext.CurrentContext.TestDirectory, "Resources/person-19900000.ldt");
                var read = reader.Read(inFile);

                Assert.That(read, Is.Not.Null);
                Assert.That(read.Count, Is.EqualTo(1));
                Assert.That(read[0], Is.Not.Null);
                Assert.That(read[0], Is.InstanceOf<libldt3.model.saetze.Auftrag>());
                Assert.That(((libldt3.model.saetze.Auftrag)read[0]).Patient, Is.Not.Null);
                Assert.That(((libldt3.model.saetze.Auftrag)read[0]).Patient.Person, Is.Not.Null);
                Assert.That(((libldt3.model.saetze.Auftrag)read[0]).Patient.Person.Geburtsdatum, Is.Not.Null);
                Assert.That(((libldt3.model.saetze.Auftrag)read[0]).Patient.Person.Geburtsdatum, Is.InstanceOf<PartialYearOnly>());
                Assert.That(((PartialYearOnly) ((libldt3.model.saetze.Auftrag)read[0]).Patient.Person.Geburtsdatum).Date, Is.EqualTo(new YearOnly(1990)));

                var writer = new LdtWriter(LdtConstants.Mode.RELAXED);
                string tempFile = Path.GetTempFileName();
                writer.Write(read, tempFile);

                Assert.That(File.ReadAllText(inFile).Trim(), Is.EqualTo(File.ReadAllText(tempFile).Trim()));
                File.Delete(tempFile);
            }

            [Test]
            public void testWithYearAnyMonth()
            {
                var reader = new LdtReader(LdtConstants.Mode.RELAXED);
                string inFile = Path.Combine(TestContext.CurrentContext.TestDirectory, "Resources/person-19900200.ldt");
                var read = reader.Read(inFile);

                Assert.That(read, Is.Not.Null);
                Assert.That(read.Count, Is.EqualTo(1));
                Assert.That(read[0], Is.Not.Null);
                Assert.That(read[0], Is.InstanceOf<libldt3.model.saetze.Auftrag>());
                Assert.That(((libldt3.model.saetze.Auftrag)read[0]).Patient, Is.Not.Null);
                Assert.That(((libldt3.model.saetze.Auftrag)read[0]).Patient.Person, Is.Not.Null);
                Assert.That(((libldt3.model.saetze.Auftrag)read[0]).Patient.Person.Geburtsdatum, Is.Not.Null);
                Assert.That(((libldt3.model.saetze.Auftrag)read[0]).Patient.Person.Geburtsdatum, Is.InstanceOf<PartialYearMonth>());
                Assert.That(((PartialYearMonth) ((libldt3.model.saetze.Auftrag)read[0]).Patient.Person.Geburtsdatum).Date, Is.EqualTo(new YearMonth(1990, 2)));

                var writer = new LdtWriter(LdtConstants.Mode.RELAXED);
                string tempFile = Path.GetTempFileName();
                writer.Write(read, tempFile);

                Assert.That(File.ReadAllText(inFile).Trim(), Is.EqualTo(File.ReadAllText(tempFile).Trim()));
                File.Delete(tempFile);
            }

            [Test]
            public void testWithFullDate()
            {
                var reader = new LdtReader(LdtConstants.Mode.RELAXED);
                string inFile = Path.Combine(TestContext.CurrentContext.TestDirectory, "Resources/person-19900204.ldt");
                var read = reader.Read(inFile);

                Assert.That(read, Is.Not.Null);
                Assert.That(read.Count, Is.EqualTo(1));
                Assert.That(read[0], Is.Not.Null);
                Assert.That(read[0], Is.InstanceOf<libldt3.model.saetze.Auftrag>());
                Assert.That(((libldt3.model.saetze.Auftrag)read[0]).Patient, Is.Not.Null);
                Assert.That(((libldt3.model.saetze.Auftrag)read[0]).Patient.Person, Is.Not.Null);
                Assert.That(((libldt3.model.saetze.Auftrag)read[0]).Patient.Person.Geburtsdatum, Is.Not.Null);
                Assert.That(((libldt3.model.saetze.Auftrag)read[0]).Patient.Person.Geburtsdatum, Is.InstanceOf<PartialLocalDate>());
                Assert.That(((PartialLocalDate) ((libldt3.model.saetze.Auftrag)read[0]).Patient.Person.Geburtsdatum).Date, Is.EqualTo(new LocalDate(1990, 2, 4)));

                var writer = new LdtWriter(LdtConstants.Mode.RELAXED);
                string tempFile = Path.GetTempFileName();
                writer.Write(read, tempFile);

                Assert.That(File.ReadAllText(inFile).Trim(), Is.EqualTo(File.ReadAllText(tempFile).Trim()));
                File.Delete(tempFile);
            }
        }
    }
}